### ISSUE TITLE
undefined method `nil_or_empty?' for String

### ISSUE THREAD

**Issue by nfrankel** [asciidoctor/jekyll-asciidoc#142]
Hello,

I'm trying to make Jekyll excerpts work, but I unfortunately come upon this trace:

```
Conversion error: Jekyll::AsciiDoc::Converter encountered an error while converting '_posts/2012-03-11-html5-still-a-non-event.adoc':
                    undefined method `nil_or_empty?' for #<String:0x007fa72f7dfdb0>
  Liquid Exception: undefined method `nil_or_empty?' for #<String:0x007fa72f7dfdb0> in tag/html-5/index.html
/Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-asciidoc-2.0.1/lib/jekyll-asciidoc/converter.rb:207:in `convert': undefined method `nil_or_empty?' for #<String:0x007fa72f7dfdb0> (NoMethodError)
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/renderer.rb:109:in `block in convert'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/renderer.rb:107:in `each'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/renderer.rb:107:in `reduce'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/renderer.rb:107:in `convert'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/renderer.rb:86:in `run'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/excerpt.rb:72:in `output'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/excerpt.rb:59:in `to_s'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/drops/document_drop.rb:22:in `excerpt'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/drops/drop.rb:52:in `public_send'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/drops/drop.rb:52:in `[]'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/context.rb:201:in `lookup_and_evaluate'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/variable_lookup.rb:48:in `block in evaluate'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/variable_lookup.rb:38:in `each_index'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/variable_lookup.rb:38:in `evaluate'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/context.rb:168:in `evaluate'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/context.rb:160:in `[]'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/condition.rb:101:in `interpret_condition'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/condition.rb:39:in `evaluate'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/tags/if.rb:38:in `block (2 levels) in render'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/tags/if.rb:37:in `each'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/tags/if.rb:37:in `block in render'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/context.rb:132:in `stack'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/tags/if.rb:36:in `render'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/block.rb:151:in `render_token'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/profiler/hooks.rb:5:in `block in render_token_with_profiling'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/profiler.rb:80:in `profile_token_render'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/profiler/hooks.rb:4:in `render_token_with_profiling'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/block.rb:135:in `block in render_all'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/block.rb:122:in `each'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/block.rb:122:in `render_all'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/block.rb:108:in `render'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/template.rb:210:in `block in render'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/template.rb:262:in `with_profiling'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/template.rb:209:in `render'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/template.rb:222:in `render!'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/liquid_renderer/file.rb:28:in `block (2 levels) in render!'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/liquid_renderer/file.rb:40:in `measure_bytes'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/liquid_renderer/file.rb:27:in `block in render!'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/liquid_renderer/file.rb:47:in `measure_time'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/liquid_renderer/file.rb:26:in `render!'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/tags/include.rb:138:in `block in render'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/context.rb:132:in `stack'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/tags/include.rb:136:in `render'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/block.rb:151:in `render_token'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/profiler/hooks.rb:5:in `block in render_token_with_profiling'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/profiler.rb:80:in `profile_token_render'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/profiler/hooks.rb:4:in `render_token_with_profiling'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/block.rb:135:in `block in render_all'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/block.rb:122:in `each'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/block.rb:122:in `render_all'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/tags/for.rb:113:in `block (2 levels) in render'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/tags/for.rb:100:in `each'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/tags/for.rb:100:in `each_with_index'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/tags/for.rb:100:in `block in render'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/context.rb:132:in `stack'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/tags/for.rb:99:in `render'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/block.rb:151:in `render_token'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/profiler/hooks.rb:5:in `block in render_token_with_profiling'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/profiler.rb:80:in `profile_token_render'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/profiler/hooks.rb:4:in `render_token_with_profiling'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/block.rb:135:in `block in render_all'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/block.rb:122:in `each'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/block.rb:122:in `render_all'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/block.rb:108:in `render'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/template.rb:210:in `block in render'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/template.rb:262:in `with_profiling'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/template.rb:209:in `render'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/liquid-3.0.6/lib/liquid/template.rb:222:in `render!'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/liquid_renderer/file.rb:28:in `block (2 levels) in render!'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/liquid_renderer/file.rb:40:in `measure_bytes'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/liquid_renderer/file.rb:27:in `block in render!'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/liquid_renderer/file.rb:47:in `measure_time'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/liquid_renderer/file.rb:26:in `render!'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/renderer.rb:134:in `render_liquid'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/renderer.rb:82:in `run'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/convertible.rb:211:in `do_layout'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/page.rb:135:in `render'
	from /Users/i303869/projects/private/nfrankel.gitlab.io/_plugins/tags.rb:43:in `write_tag_index'
	from /Users/i303869/projects/private/nfrankel.gitlab.io/_plugins/tags.rb:36:in `block in generate'
	from /Users/i303869/projects/private/nfrankel.gitlab.io/_plugins/tags.rb:35:in `each'
	from /Users/i303869/projects/private/nfrankel.gitlab.io/_plugins/tags.rb:35:in `generate'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/site.rb:174:in `block in generate'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/site.rb:172:in `each'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/site.rb:172:in `generate'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/site.rb:68:in `process'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/command.rb:26:in `process_site'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/commands/build.rb:63:in `build'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/commands/build.rb:34:in `process'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/lib/jekyll/commands/serve.rb:37:in `block (2 levels) in init_with_program'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `call'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `block in execute'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `each'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `execute'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/mercenary-0.3.6/lib/mercenary/program.rb:42:in `go'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/mercenary-0.3.6/lib/mercenary.rb:19:in `program'
	from /Users/i303869/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jekyll-3.3.1/exe/jekyll:13:in `<top (required)>'
	from /usr/local/bin/jekyll:22:in `load'
	from /usr/local/bin/jekyll:22:in `<main>'
```

I'm using:

* Ruby 2.2.3
* Jekyll 3.3.1
* jekyll-asciidoctor plugin 2.0.1

Interestingly enough, it works if:

* Either I remove *.adoc files
* Or don't try to write excerpts

Hints welcome, many thanks.

**Comment by mojavelinux** [comment]
I'll need to investigate to see how the call chain is arriving at this point. However, what I can tell you is that the excerpts are extremely problematic, esp when used with this plugin. In my opinion, they are very poorly done and hard to accommodate in extensions. Filters don't get applied to them, and there is no way to know what document they came from. Therefore, we don't have a good opportunity to prepare them properly. My recommendation is to avoid using them and instead have your own way of doing excerpts. We may end up providing an alternative excerpt key in this plugin.

**Comment by nfrankel** [comment]
Thanks for your feedback. So far, I'm living without excerpts and I've no clue how they are coded - I'm not a Ruby developer, thus I believe you. However, from a functional point of view, my home page is very long and would benefit from excerpts.

Whatever you can provide is good enough for me.

**Comment by theaspect** [comment]
Same issue for me

**Comment by mojavelinux** [comment]
Try as I might, I cannot reproduce this. What I am trying to figure out is how the object being passed to convert is not a String. If it's a String, it should have the `nil_or_empty?` method because it is added to String by Asciidoctor. However, in the event that it is not a String, it makes sense why it would not find this method. I just can't figure out how that's happening.

Can you please share a site that demonstrates this problem so that I may study it? I know that switching from `nil_or_empty?` would fix it, but it would also mean that we cannot assume that we're getting a String. And I need to know what that's all about before I make this fix.

**Comment by mojavelinux** [comment]
> However, from a functional point of view, my home page is very long and would benefit from excerpts.

Could you provide more details about how you are defining your excerpts? Are you defining them using page-excerpt or are you depending on the auto-extract feature? That feature definitely still needs work in the plugin...if I can wrap my head around the mess that Jekyll has made.

**Comment by nfrankel** [comment]
I tried again. I didn't do configure anything so that the excerpt is the default one (probably the body itself). Then in my layout, I call `{post.excerpt}`.

But I pinpointed the cause of the error in a Ruby plugin. If I remove the plugin, it works. If I remove some tags for some posts, it works also. I don't know why, I have 0 in Ruby knowledge.

```ruby
#From https://github.com/arocks/arunrocks
module Jekyll

  class TagIndex < Page
    def initialize(site, base, dir, tag)
      @site = site
      @base = base
      @dir = dir
      @name = 'index.html'
      self.process(@name)
      self.read_yaml(File.join(base, '_layouts'), 'tag.html')
      self.data['tag'] = tag
Jekyll.logger.info "tag #{tag}"
      self.data['related'] = []
      site.tags[tag].each do |post|
        post.data['tags'].each do |rel| 
          self.data['related'].push(rel)
        end
      end
      self.data['related'] = self.data['related'].uniq
      tag_title_prefix = site.config['tag_title_prefix'] || 'Tags: '
      self.data['title'] = "#{tag_title_prefix}#{tag}"
      self.data['sitemap'] = true
    end
  end

  class TagGenerator < Generator
    safe true
    priority :high
    
    def generate(site)
      if site.layouts.key? 'tag'
        dir = site.config['tag_dir'] || 'tag'
        site.tags.keys.each do |tag|
          write_tag_index(site, File.join(dir, tag.gsub(/\s/, "-").gsub(/[^\w-]/, '').downcase), tag)
        end
      end
    end
  
    def write_tag_index(site, dir, tag)
      index = TagIndex.new(site, site.source, dir, tag)
      index.render(site.layouts, site.site_payload)
      index.write(site.dest)
      site.static_files << index
    end
  end
end
```

**Comment by mojavelinux** [comment]
As I said, I understand why it's happening. It's an ordering issue. I just don't know what exactly is triggering it (aside from generally this plugin, which we already suspect).

Is there any way you can share a minimal project that fails in the same way? (or give me access to your site's repository?) There is a complex interaction between plugins that I need to understand in order to know how to fix it.

**Comment by mojavelinux** [comment]
I also think I understand now how to make excerpts work correctly...or at least I have an idea. But I need to test that theory.

**Comment by nfrankel** [comment]
My project is hosted on Gitlab. Please create an account, give me its handle and I'll give you access.

**Comment by mojavelinux** [comment]
My account on GitLab is mojavelinux also. (https://gitlab.com/mojavelinux)

**Comment by nfrankel** [comment]
I granted you access rights. I suggest you use branch `features/new-posts` because `master` has a lot of posts and generation takes a lot of time.

**Comment by mojavelinux** [comment]
Great! I'll have a look.

**Comment by DanielSWolf** [comment]
Any news on this? I'm trying to write an AsciiDoctor plugin, and it's failing as soon as my class calls `use_dsl`, because `use_dsl` executes `self.name.nil_or_empty?`.

> jekyll 3.5.1 | Error:  undefined method `nil_or_empty?' for "CustomAdmonitionBlock":String

At that point, `self.name` appears to be a String, but it doesn't have a method `nil_or_empty`. To be sure, I added the following debugging code:

```rb
class CustomAdmonitionBlock < Asciidoctor::Extensions::BlockProcessor
  puts 'value: ' + self.name
  puts 'class: ' + self.name.class.to_s
  puts 'methods: ' + self.name.methods.sort.join(', ')
  puts 'responds to `nil_or_empty?`: ' + self.name.respond_to?(:nil_or_empty?).to_s
end
```

Output:
> value: CustomAdmonitionBlock
> class: String
> methods: !, !=, !~, %, *, +, +@, -@, <, <<, <=, <=>, ==, ===, =~, >, >=, [], []=, __id__, __send__, ansi_jump, ascii_only?, b, between?, black, blue, bold, bytes, bytesize, byteslice, capitalize, capitalize!, casecmp, center, chars, chomp, chomp!, chop, chop!, chr, class, clear, clear_line, clear_screen, clone, codepoints, colorize, concat, count, crypt, cyan, define_singleton_method, delete, delete!, display, downcase, downcase!, dump, dup, each_byte, each_char, each_codepoint, each_line, empty?, encode, encode!, encoding, end_with?, enum_for, eql?, equal?, extend, force_encoding, freeze, frozen?, gem, getbyte, green, gsub, gsub!, has_ansi?, has_color?, hash, hex, include?, index, insert, inspect, instance_eval, instance_exec, instance_of?, instance_variable_defined?, instance_variable_get, instance_variable_set, instance_variables, intern, is_a?, itself, kind_of?, length, lines, ljust, lstrip, lstrip!, magenta, match, method, methods, next, next!, nil?, object_id, oct, ord, parse_csv, partition, prepend, private_methods, protected_methods, psych_to_yaml, public_method, public_methods, public_send, red, remove_instance_variable, replace, reset_ansi, reset_color, respond_to?, reverse, reverse!, rindex, rjust, rpartition, rstrip, rstrip!, scan, scrub, scrub!, send, setbyte, singleton_class, singleton_method, singleton_methods, size, slice, slice!, split, squeeze, squeeze!, start_with?, strip, strip!, strip_ansi, strip_color, sub, sub!, succ, succ!, sum, swapcase, swapcase!, taint, tainted?, tap, to_c, to_enum, to_f, to_i, to_json, to_json_raw, to_json_raw_object, to_liquid, to_r, to_s, to_str, to_sym, to_yaml, to_yaml_properties, tr, tr!, tr_s, tr_s!, trust, unicode_normalize, unicode_normalize!, unicode_normalized?, unpack, untaint, untrust, untrusted?, upcase, upcase!, upto, valid_encoding?, white, yellow
> responds to `nil_or_empty?`: false

**Comment by mojavelinux** [comment]
Until I can reproduce this problem in a sample project or test environment, I can't fix it because I don't know what code to change. As soon as someone provides a sample project I can use to test, then I can try to resolve it.

@nfrankel did provide me access to his site build, but I haven't had a chance to try it yet. But what would speed this up is if someone provided a public sample project that demonstrates the problem without a lot of other content or configuration in it.

**Comment by mojavelinux** [comment]
I finally understand what the problem is. Under certain conditions, the very first call to the Jekyll AsciiDoc is the `convert` method on `Converter` class. This can happen if a custom generator is converting pages, such as the tags plugin.

At this time, the plugin may not be fully initialized, meaning the `setup` method hasn't yet been called. The `setup` method is what loads Asciidoctor. So basically, there's an attempt to use a method provided by Asciidoctor before Asciidoctor is loaded.

The fix is to switch the order of calls so that setup is called first, then `content.nil_or_empty?`.

**Comment by mojavelinux** [comment]
@DanielSWolf To fix your problem, you just need to require asciidoctor in your extensions file.

```ruby
require 'asciidoctor'
require 'asciidoctor/extensions'

class CustomAdmonitionBlock < Asciidoctor::Extensions::BlockProcessor
  use_dsl
end
```

This plugin loads Asciidoctor lazily, so you shouldn't assume it is already loaded.

Though, this does make me relies that requiring `asciidoctor/extensions` should fail if Asciidoctor is not yet loaded.

**Comment by mojavelinux** [comment]
I realize now why I couldn't reproduce. I was loading asciidoctor-diagram, which was in turn loading Asciidoctor.

There are at least two ways to workaround the problem before I apply the patch.

1. Add Asciidoctor to the `:jekyll_plugins` group in your Gemfile:

```
group :jekyll_plugins do
  gem 'asciidoctor'
end
```

2. Create a blank plugin file at _plugins/asciidoctor.rb that requires Asciidoctor

```
require 'asciidoctor'
```

Long story, short, you need to force Asciidoctor to load eagerly to avoid this sort of problem.

**Comment by nfrankel** [comment]
Thanks for the answer. I'll try this!

**Comment by mojavelinux** [comment]
@nfrankel I just tested with your blog and it fixes the error. I just added the following to the top of _plugins/tags.rb

```
require 'asciidoctor'
```

One thing I noticed about that tags plugin is that it runs before the Jekyll AsciiDoc integrator runs. Therefore, the only metadata available at that time is metadata defined in the front matter. The information in the AsciiDoc document header is not yet seen. But that's okay in your case since you are defining all metadata in front matter. But it will be something we'll have to sort out if others want to use the tags plugin with integrated AsciiDoc metadata.
